src/transaction/confirm.rs

```diff
@@ -79,228 +70,190 @@ impl Default for ConfirmParams {
                     )
                 }
-                ConfirmState::PollCheckingBlockNumber(block_num, target_block_num) => {
-                    let block_num = match ready!(block_num.poll(cx)) {
-                        Ok(block_num) => block_num,
-                        Err(err) => return Poll::Ready(Err(err.into())),
-                    };
-
-                    if block_num >= *target_block_num {
-                        ConfirmState::Checking(future::try_join(
-                            MaybeReady::ready(Ok(block_num)),
-                            web3.eth().transaction_receipt(*tx).compat(),
-                        ))
-                    } else {
-                        ConfirmState::PollDelay(delay(params.poll_interval), *target_block_num)
-                    }
-                }
             };
 
-            self.state = next_state;
+        if let Some(block_timeout) = self.params.block_timeout {
+            let starting_block = *self.starting_block.get_or_insert(latest_block);
+            let elapsed_blocks = latest_block.saturating_sub(starting_block);
+
+            if elapsed_blocks > U64::from(block_timeout) {
+                return Err(ExecutionError::ConfirmTimeout);
+            }
         }
+
+        Ok(Check::Pending {
+            target_block,
+            remaining_confirmations,
+        })
     }
-}
 
-impl<T: Transport> Debug for ConfirmState<T> {
-    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
-        match self {
-            ConfirmState::Check => f.debug_tuple("Check").finish(),
-            ConfirmState::Checking(_) => f.debug_tuple("Checking").finish(),
-            ConfirmState::CreatingFilter(_, t, c) => {
-                f.debug_tuple("CreatingFilter").field(t).field(c).finish()
-            }
-            ConfirmState::WaitingForBlocks(_, t) => {
-                f.debug_tuple("WaitingForBlocks").field(t).finish()
-            }
-            ConfirmState::PollDelay(d, t) => f.debug_tuple("PollDelay").field(d).field(t).finish(),
-            ConfirmState::PollCheckingBlockNumber(_, t) => {
-                f.debug_tuple("PollCheckingBlockNumber").field(t).finish()
+    /// Waits for blocks to be mined. This method tries to use a block filter to
+    /// wait for a certain number of blocks to be mined. If that fails, it falls
+    /// back to polling the latest block number to wait until a target block
+    /// number is reached.
+    ///
+    /// This method returns the latest block number if it is known. Specifically
+    /// if the polling method is used to query the latest block, then this will
+    /// return the `target_block` since the node is currently at that block
+    /// height. This method returns `None` otherwise as block filters can emit
+    /// block hashes for blocks at the same height because of re-orgs, so the
+    /// latest block needs to be queried after the waiting period as it is not
+    /// garanteed to be `target_block`.
+    async fn wait_for_blocks(
```

Comment:
This method can be refactored a bit. Specifically, we can have the `wait_for_blocks_with_filter` check that the target block is actually reached and not have to query the transaction receipt needlessly.

I plan on doing this in a follow up PR though, as I wanted the async implementation to be equivalent to the previous state machine one for now.

Reply:
Now that I think about this more - I think it should stay as is. The reason being if there is a reorg, then the transaction receipt might have been invalidated so its worth re-querying to make re-calculate the remaining blocks.

Reply:
But what are the chances that the reorg will happen between your `wait_for_blocks_with_filter` finishes and `check` re-executes? I feel like the reorg risk will always be there and should be "priced-in" to the confirmation threshold passed.

Reply:
Sorry I didn't explain it very well.

So the issue with how the code works now, is that `eth_newBlockFilter` and `eth_getFilterChanges` emits a new block hash for each new block. If a block `X` gets mined at height `n`, then a re-org happens and `X` gets removed and instead another block `Y` takes its place at height `n`, then we will potentially receive 2 block hashes with `eth_getFilterChanges`. This makes the `wait_for_blocks_with_filter` code potentially return *before* target block was reached.

In the current implementation (and previous one as well) used to just re-check unconditionally after `n` new block hashes were received by re-querying the current latest block and the transaction receipt, and check to see if enough confirmations happen. If not, we would create a new block filter and the process would restart.

I proposed changing the `wait_for_blocks_with_filter` implementation (making behave differently than the old state-machine based implementation) so that after receiving `remaining_confirmation` new blocks, we would check with the current latest block and continue polling the block filter if we hadn't reached the `target_block`. This has the advantage of potentially querying the transaction receipt one less time and not needlessly creating a new block filter. However, it also could potentially cause the future to wait for an additional block, in the case that the reorg cause the transaction to move from some block `n` to `n-1`.

I hope this clarifies it a bit better. Maybe take a look at #325 which I implement this alternate strategy - it might make a littler more sense then.

I think one thing we should do is reuse the block filter. It was very hard previously in the state machine implementation (implementing loops was tedious...) but is much easier now.

Reply:
Yes I see now and agree that we should check the blocknumber of the most recent block. I like the idea of doing this inside `wait_for_blocks_with_filter` so that we don't re-query the block number in case we are using the polling implementation (but at the end of the day that's a really minor improvement).

Reply:
> we don't re-query the block number in case we are using the polling implementation

This is already the case.